### DISCUSSION TITLE
How to scrape page follower's/liker's information

### DISCUSSION THREAD

**Discussion by Gardek9** [kevinzg/facebook-scraper#373] [Q&A]
Hello,

I just discovered your scraper and it may be just what I need. Can it scrape user name, user id, first name, last name, email (if public), phone number (if public) of ALL the followers and/or likers of a page and/or group?

If so, can you walk me on how it would be? How to setup tor (on linux mint 19.3) so I don’t get banned and then dump the results on a csv file?

Thank you!

**Comment by neon-ninja** [comment]
No, Facebook doesn't make this information available for privacy reasons.

**Reply by Gardek9** [reply]
Can it scrape the comments and replies of a page and/or group and thus identify the commenter's and replier's user name and user id? If so, what would be the setup? Thanks

**Reply by neon-ninja** [reply]
Yes. You would pass the `options={"comments": True}` option to `get_posts`. Here's an example:

```python
from pprint import pprint
from facebook_scraper import *
import logging
import os
import json

# enable_logging(logging.DEBUG)

set_cookies("cookies.txt")
post = next(get_posts("Nintendo", options={"comments": True, "progress": True}))
pprint(post)
```
output:

```python
{'available': True,
 'comments': 106,
 'comments_full': [{'comment_id': '4271137392970681',
                    'comment_image': None,
                    'comment_text': 'Got a Facebook ad for these and raced '
                                    'over to the online store...\n'
                                    '\n'
                                    'Not every design is available in every '
                                    "cut, don't be surprised when the one you "
                                    "like from this ad doesn't exist as a "
                                    "men's or a women's or a kid's size.\n"
                                    '\n'
                                    "Couldn't you guys go with someone like "
                                    'Teespring that prints on demand? All '
                                    'sizes and colors always available.',
                    'comment_time': datetime.datetime(2021, 7, 2, 6, 8, 51),
                    'comment_url': 'https://facebook.com/4271137392970681',
                    'commenter_id': '17214041',
                    'commenter_meta': None,
                    'commenter_name': 'Alexa John Danielson',
                    'commenter_url': 'https://facebook.com/Desirsar?fref=nf&rc=p&__tn__=R'},
```
and so on.

Then you could pass commenter ids to `get_profile` like so:

```python
pprint(get_profile('17214041'))
```
output:
```python
{'About': "I play guitar and bass, I kinda stopped pretending I'm good at "
          'sports anymore (no one wants me on a team, anyway), but I still run '
          'or practice sports when I can... umm... not sure what to put here '
          "anymore, I really don't like the person I am and have been trying "
          "to change, but it's harder than I thought it would be.",
 'Basic Info': 'September 19, 1978\n'
               'Birthday\n'
               'Female\n'
               'Gender\n'
               'Men and Women\n'
               'Interested In\n'
               'American English, English English, Australian English, '
               'Japanese language and Spanish language\n'
               'Languages\n'
               'Discordian Society\n'
               'Religious Views\n'
               'Libertarian Party\n'
               'Political Views',
 'Contact Info': {'Email': 'clu@inebraska.com',
                  'Facebook': '/Desirsar',
                  'Instagram': 'Desirsar',
                  'Mobile': '+1 913-593-3240',
                  'Snapchat': 'Desirsar',
                  'SoundCloud': 'Desirsar',
                  'Twitch': 'desirsar1',
                  'Twitter': 'Desirsar',
                  'Windows Live Messenger': 'clu@inebraska.com'},
 'Education': 'University of Nebraska–Lincoln\n'
              'Communication Studies · Spanish language · Japanese (language)\n'
              'September 2006 - August 2011\n'
              'University of Nebraska-Lincoln\n'
              'Broadcasting · Spanish · Japanese\n'
              'September 2002 - August 2005\n'
              'Lincoln Northeast High School\n'
              'High School\n'
              'Class of 1996',
 'Family Members': 'Dawn Danielson Lomax\n'
                   'Sister\n'
                   'Dayna Danielson Hokinson\n'
                   'Sister\n'
                   'DiAnna Danielson\n'
                   'Sister\n'
                   'Dee Dee Danielson Reinke\n'
                   'Sister\n'
                   'Maggie Gallagher Danielson\n'
                   'Mother\n'
                   'Don Danielson\n'
                   'Cousin\n'
                   'Benjamin Charles Houdesheldt\n'
                   'Cousin\n'
                   'Joe Houdesheldt\n'
                   'Cousin\n'
                   'Mike Houdesheldt\n'
                   'Cousin\n'
                   'Haley Staack\n'
                   'Cousin\n'
                   'Karissa Fortney\n'
                   'Cousin\n'
                   'Anita Moore\n'
                   'Cousin\n'
                   'Ashley Gallagher\n'
                   'Cousin\n'
                   'Meghan Rose Gallagher\n'
                   'Cousin',
 'Life Events': '2021\n'
                'Started New Job at Nelnet\n'
                "Left Job at Dunkin'\n"
                '2019\n'
                "Started New Job at Dunkin'\n"
                '2017\n'
                'Moved to Lincoln, Nebraska, USA\n'
                '2011\n'
                'Left University of Nebraska–Lincoln\n'
                '2006\n'
                'Started School at University of Nebraska–Lincoln\n'
                '2005\n'
                'Left University of Nebraska-Lincoln\n'
                '2002\n'
                'Started School at University of Nebraska-Lincoln',
 'Name': 'Alexa John Danielson',
 'Places Lived': 'Lincoln, Nebraska\n'
                 'Current City\n'
                 'Lincoln, Nebraska, USA\n'
                 'Moved in 2017\n'
                 'Lincoln, Nebraska\n'
                 'Hometown',
 'Relationship': 'Single',
 'Work': 'Nelnet\n'
         'April 2021 - Present\n'
         'Lincoln, Nebraska\n'
         "Dunkin'\n"
         'August 18, 2019 - April 2021\n'
         'Lincoln, Nebraska',
 'cover_photo': 'https://scontent.fhlz2-1.fna.fbcdn.net/v/t1.6435-9/fr/cp0/e15/q65/38838045_10104092982397433_6182194939682619392_n.jpg?_nc_cat=101&_nc_rgb565=1&ccb=1-3&_nc_sid=dd9801&efg=eyJpIjoidCJ9&_nc_ohc=w1qzmhCiYToAX99iDee&_nc_ht=scontent.fhlz2-1.fna&tp=14&oh=c54a413041da7bdce1d835b388425ae0&oe=60E3F3B0&manual_redirect=1',
 'cover_photo_text': "Cover Photo: Alexa John Danielson's photo.",
 'id': '17214041',
 'profile_picture': 'https://scontent.fhlz2-1.fna.fbcdn.net/v/t1.6435-9/fr/cp0/e15/q65/95438274_10105421951402893_8345971747983458304_n.jpg?_nc_cat=111&_nc_rgb565=1&ccb=1-3&_nc_sid=85a577&efg=eyJpIjoidCJ9&_nc_ohc=8_I8tpyb1bkAX_j3Li7&_nc_ht=scontent.fhlz2-1.fna&tp=14&oh=d04c99ad61f2846539034f001f34ddb2&oe=60E417EC&manual_redirect=1'}
```
Also I wouldn't recommend using tor unless you know what you're doing, as Facebook has a tendency to block tor users.

**Reply by Gardek9** [reply]
Ok, great. So now how do I tell it to dump the commenter_id, commenter_name and commenter_url to a csv file? There will be many.
Then, for the second part, how do I do a bulk profile print of all the commenter_ids and then have it dump their email, name, phone number, etc. into a new csv file so I can make sense of it all?
Then how do I run a list of proxies (you said Tor would be a bad idea)? Thank so much.

**Reply by Gardek9** [reply]
In the meantime I gave your example a try and a got a cookies.txt not found error, a locale error (if I do locale.getlocale(), I get ('en_US', 'UTF-8'), but when I do post = next......, I get locale is set to es_LA, switch to en_US for better results).
After doing all this I get only ONE post from the page owner (Nintendo {mine is different}).
You wouldn't be open to setting up the scarper to do what I need it to, would you? I am trying to build a community and I need to connect with them BEFORE my launch because they are the ones who will provide the content through their community interactions. Thanks again in advance!

**Reply by neon-ninja** [reply]
I like to use the pandas library for this kind of thing, as it makes writing to a CSV a 1-2 line operation. Here's how I would do this:

```python
#!/usr/bin/env python
from pprint import pprint
from facebook_scraper import *
import logging
import os
import json
import pandas as pd
from tqdm import tqdm
import time

# enable_logging(logging.DEBUG)

set_cookies("cookies.json")
commenters = {}
for post in get_posts("Nintendo", options={"comments": True, "progress": True}, pages=10):
    if post.get("comments_full"):
        for c in post.get("comments_full"):
            commenters[c.get("commenter_id")] = c.get("commenter_name")
            for r in c.get("replies", []):
                commenters[r.get("commenter_id")] = r.get("commenter_name")
df = pd.DataFrame(commenters.items(), columns=["id", "name"])
df.to_csv("commenters.csv", index=False)

profiles = []
for user_id in tqdm(commenters.keys()):
    while True:
        try:
            profile = get_profile(user_id, timeout=60)
            profiles.append(profile)
            break
        except exceptions.TemporarilyBanned:
            print("TemporarilyBanned, sleeping 10m")
            time.sleep(600)
        except Exception as e:
            print(f"Unable to get profile {user_id} due to {type(e)}:{e}")
            break
    time.sleep(1)
df = pd.DataFrame(profiles)
df.to_csv("profiles.csv", index=False)
```

I wouldn't recommend running a list of proxies unless you know what you're doing either, as Facebook has a tendency to block proxies. Note that you don't need the commenter URL, as you can use the profile ID to access anyone's profile page. For example, facebook.com/17214041
You need to extract your cookies as per the README. If your Facebook language is set to English, that should solve the locale warning.
You only got one post as I used the `next` function to get the first result from the get_posts generator, as an example.

**Reply by Gardek9** [reply]
Ok, I will give a go, but first, I exported the Facebook cookie using cookie-editor to a cookies.txt file, except that I don't know what to do with it or where to put it. Also I noticed that before you set cookies to a cookies.txt file and now you set it a cookies.json, how do I handle this? Thanks again!

**Reply by Gardek9** [reply]
I am getting a syntax error when I type the following line:
df = pd.DataFrame(commenters.items(), columns=["id", "name"])
it points to the df part.
Also I created a fbspider.py file to put all that code in. I placed it in the facebook_scraper folder, but I get a 'fbspider' not defined error. I'm on a HUGE learning curve here, the thing is, I am not a programmer, but I do need to get this done and I can follow direction. Sorry for not knowing enough, but I'm trying. Thanks.

**Reply by neon-ninja** [reply]
When you use the `set_cookies` function, you give it the path to your cookies file. This path can be relative to where you're running the script from. Either Netscape format or JSON format should work fine.

Instead of retyping all of that code, why not copy paste it? The SyntaxError is likely due to a typo of yours.

You would run this script like so `python fbspider.py`

**Reply by Gardek9** [reply]
Ok, so I have my fspider.py (I named it fspider.py NOT fbspider.py) file (which I copied and pasted from your code above) in the facebook_scraper folder along with my cookies.json file (same folder).
If I run python fspider.py I get a syntax error in the print(f"Unable to get profile {user_id} due to {type(e)}:{e}") points to the last " (remember I did a copy and paste of your code).
If I run python3 fspider.py I get:
/usr/lib/python3/dist-packages/requests/__init__.py:80: RequestsDependencyWarning: urllib3 (1.26.6) or chardet (3.0.4) doesn't match a supported version!
  RequestsDependencyWarning)
Traceback (most recent call last):
  File "fspider.py", line 3, in <module>
    from facebook_scraper import *
  File "/home/laptop/.local/lib/python3.6/site-packages/facebook_scraper/facebook_scraper.py", line 13, in <module>
    from . import utils
ImportError: attempted relative import with no known parent package

**Reply by neon-ninja** [reply]
fspider.py should not be in the same folder as facebook_scraper.py. Move it one level up at least. Or somewhere else in your filesystem. `python` on its own probably is python 2 on your system in that case.

**Reply by Gardek9** [reply]
Ok, It seems to work, however I get zero results (I ran your Nintendo example exactly as you wrote it just to see it work, but I got zero results).

**Reply by neon-ninja** [reply]
Uncomment `enable_logging(logging.DEBUG)` and see if you get any error logs

**Reply by Gardek9** [reply]
Hi, Here is the log file. Apparently, there are no comments.
The pages that I want to scrape have millions of followers (one has almost 8M, another almost 4M and a bunch of them have around 1M each). I noticed this only scrapes 10 pages and obviously my pages should have millions of commenters. How do I set it up so I get them all (if possible). Also, I am focusing on  commenters based in the United States for now. Thanks

[fspider_py_log.txt](https://github.com/kevinzg/facebook-scraper/files/6759902/fspider_py_log.txt)

**Reply by neon-ninja** [reply]
Try update to the latest master branch. Increase `pages=10` to your desired page limit. Or set it to `None` to try retrieve all posts ever made by the page (might take a long time)

**Reply by Gardek9** [reply]
I tried pip3 install git+https://github.com/kevinzg/facebook-scraper.git, but got:

$ pip3 install git+https://github.com/kevinzg/facebook-scraper.git
Defaulting to user installation because normal site-packages is not writeable
Collecting git+https://github.com/kevinzg/facebook-scraper.git
  Cloning https://github.com/kevinzg/facebook-scraper.git to /tmp/pip-req-build-pxw3v0ha
  Running command git clone -q https://github.com/kevinzg/facebook-scraper.git /tmp/pip-req-build-pxw3v0ha
  ERROR: Error [Errno 2] No such file or directory: 'git': 'git' while executing command git clone -q https://github.com/kevinzg/facebook-scraper.git /tmp/pip-req-build-pxw3v0ha
ERROR: Cannot find command 'git' - do you have 'git' installed and in your PATH?

I tried to do pip3 install git and got:

Defaulting to user installation because normal site-packages is not writeable
ERROR: Could not find a version that satisfies the requirement git (from versions: none)
ERROR: No matching distribution found for git

And if I do pip3 install facebook-scraper -U, I get:

Defaulting to user installation because normal site-packages is not writeable
Requirement already satisfied: facebook-scraper in ./.local/lib/python3.6/site-packages (0.2.43)
Requirement already satisfied: demjson<3.0.0,>=2.2.4 in ./.local/lib/python3.6/site-packages (from facebook-scraper) (2.2.4)
Requirement already satisfied: requests-html<0.11.0,>=0.10.0 in ./.local/lib/python3.6/site-packages (from facebook-scraper) (0.10.0)
Requirement already satisfied: dateparser<2.0.0,>=1.0.0 in ./.local/lib/python3.6/site-packages (from facebook-scraper) (1.0.0)
Requirement already satisfied: regex!=2019.02.19 in ./.local/lib/python3.6/site-packages (from dateparser<2.0.0,>=1.0.0->facebook-scraper) (2021.4.4)
Requirement already satisfied: python-dateutil in ./.local/lib/python3.6/site-packages (from dateparser<2.0.0,>=1.0.0->facebook-scraper) (2.8.1)
Requirement already satisfied: pytz in /usr/lib/python3/dist-packages (from dateparser<2.0.0,>=1.0.0->facebook-scraper) (2018.3)
Requirement already satisfied: tzlocal in ./.local/lib/python3.6/site-packages (from dateparser<2.0.0,>=1.0.0->facebook-scraper) (2.1)
Requirement already satisfied: requests in /usr/lib/python3/dist-packages (from requests-html<0.11.0,>=0.10.0->facebook-scraper) (2.18.4)
Requirement already satisfied: parse in ./.local/lib/python3.6/site-packages (from requests-html<0.11.0,>=0.10.0->facebook-scraper) (1.19.0)
Requirement already satisfied: fake-useragent in ./.local/lib/python3.6/site-packages (from requests-html<0.11.0,>=0.10.0->facebook-scraper) (0.1.11)
Requirement already satisfied: pyquery in ./.local/lib/python3.6/site-packages (from requests-html<0.11.0,>=0.10.0->facebook-scraper) (1.4.3)
Requirement already satisfied: pyppeteer>=0.0.14 in ./.local/lib/python3.6/site-packages (from requests-html<0.11.0,>=0.10.0->facebook-scraper) (0.2.5)
Requirement already satisfied: w3lib in ./.local/lib/python3.6/site-packages (from requests-html<0.11.0,>=0.10.0->facebook-scraper) (1.22.0)
Requirement already satisfied: bs4 in ./.local/lib/python3.6/site-packages (from requests-html<0.11.0,>=0.10.0->facebook-scraper) (0.0.1)
Requirement already satisfied: tqdm<5.0.0,>=4.42.1 in ./.local/lib/python3.6/site-packages (from pyppeteer>=0.0.14->requests-html<0.11.0,>=0.10.0->facebook-scraper) (4.61.1)
Requirement already satisfied: appdirs<2.0.0,>=1.4.3 in ./.local/lib/python3.6/site-packages (from pyppeteer>=0.0.14->requests-html<0.11.0,>=0.10.0->facebook-scraper) (1.4.4)
Requirement already satisfied: importlib-metadata<3.0.0,>=2.1.1 in ./.local/lib/python3.6/site-packages (from pyppeteer>=0.0.14->requests-html<0.11.0,>=0.10.0->facebook-scraper) (2.1.1)
Requirement already satisfied: pyee<9.0.0,>=8.1.0 in ./.local/lib/python3.6/site-packages (from pyppeteer>=0.0.14->requests-html<0.11.0,>=0.10.0->facebook-scraper) (8.1.0)
Requirement already satisfied: urllib3<2.0.0,>=1.25.8 in ./.local/lib/python3.6/site-packages (from pyppeteer>=0.0.14->requests-html<0.11.0,>=0.10.0->facebook-scraper) (1.26.6)

**Reply by Gardek9** [reply]
I'm doing a sudo apt install git-all and it seems to be installing git right now

**Reply by Gardek9** [reply]
And after installing git, I still get:

$ pip3 install git+https://github.com/kevinzg/facebook-scraper.git
Defaulting to user installation because normal site-packages is not writeable
Collecting git+https://github.com/kevinzg/facebook-scraper.git
  Cloning https://github.com/kevinzg/facebook-scraper.git to /tmp/pip-req-build-0vqbmf9e
  Running command git clone -q https://github.com/kevinzg/facebook-scraper.git /tmp/pip-req-build-0vqbmf9e
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
    Preparing wheel metadata ... done
Requirement already satisfied: demjson<3.0.0,>=2.2.4 in ./.local/lib/python3.6/site-packages (from facebook-scraper==0.2.43) (2.2.4)
Requirement already satisfied: requests-html<0.11.0,>=0.10.0 in ./.local/lib/python3.6/site-packages (from facebook-scraper==0.2.43) (0.10.0)
Requirement already satisfied: dateparser<2.0.0,>=1.0.0 in ./.local/lib/python3.6/site-packages (from facebook-scraper==0.2.43) (1.0.0)
Requirement already satisfied: python-dateutil in ./.local/lib/python3.6/site-packages (from dateparser<2.0.0,>=1.0.0->facebook-scraper==0.2.43) (2.8.1)
Requirement already satisfied: tzlocal in ./.local/lib/python3.6/site-packages (from dateparser<2.0.0,>=1.0.0->facebook-scraper==0.2.43) (2.1)
Requirement already satisfied: regex!=2019.02.19 in ./.local/lib/python3.6/site-packages (from dateparser<2.0.0,>=1.0.0->facebook-scraper==0.2.43) (2021.4.4)
Requirement already satisfied: pytz in /usr/lib/python3/dist-packages (from dateparser<2.0.0,>=1.0.0->facebook-scraper==0.2.43) (2018.3)
Requirement already satisfied: parse in ./.local/lib/python3.6/site-packages (from requests-html<0.11.0,>=0.10.0->facebook-scraper==0.2.43) (1.19.0)
Requirement already satisfied: pyppeteer>=0.0.14 in ./.local/lib/python3.6/site-packages (from requests-html<0.11.0,>=0.10.0->facebook-scraper==0.2.43) (0.2.5)
Requirement already satisfied: requests in /usr/lib/python3/dist-packages (from requests-html<0.11.0,>=0.10.0->facebook-scraper==0.2.43) (2.18.4)
Requirement already satisfied: pyquery in ./.local/lib/python3.6/site-packages (from requests-html<0.11.0,>=0.10.0->facebook-scraper==0.2.43) (1.4.3)
Requirement already satisfied: w3lib in ./.local/lib/python3.6/site-packages (from requests-html<0.11.0,>=0.10.0->facebook-scraper==0.2.43) (1.22.0)
Requirement already satisfied: fake-useragent in ./.local/lib/python3.6/site-packages (from requests-html<0.11.0,>=0.10.0->facebook-scraper==0.2.43) (0.1.11)
Requirement already satisfied: bs4 in ./.local/lib/python3.6/site-packages (from requests-html<0.11.0,>=0.10.0->facebook-scraper==0.2.43) (0.0.1)
Requirement already satisfied: websockets<9.0,>=8.1 in ./.local/lib/python3.6/site-packages (from pyppeteer>=0.0.14->requests-html<0.11.0,>=0.10.0->facebook-scraper==0.2.43) (8.1)
Requirement already satisfied: urllib3<2.0.0,>=1.25.8 in ./.local/lib/python3.6/site-packages (from pyppeteer>=0.0.14->requests-html<0.11.0,>=0.10.0->facebook-scraper==0.2.43) (1.26.6)
Requirement already satisfied: pyee<9.0.0,>=8.1.0 in ./.local/lib/python3.6/site-packages (from pyppeteer>=0.0.14->requests-html<0.11.0,>=0.10.0->facebook-scraper==0.2.43) (8.1.0)
Requirement already satisfied: importlib-metadata<3.0.0,>=2.1.1 in ./.local/lib/python3.6/site-packages (from pyppeteer>=0.0.14->requests-html<0.11.0,>=0.10.0->facebook-scraper==0.2.43) (2.1.1)
Requirement already satisfied: tqdm<5.0.0,>=4.42.1 in ./.local/lib/python3.6/site-packages (from pyppeteer>=0.0.14->requests-html<0.11.0,>=0.10.0->facebook-scraper==0.2.43) (4.61.1)
Requirement already satisfied: appdirs<2.0.0,>=1.4.3 in ./.local/lib/python3.6/site-packages (from pyppeteer>=0.0.14->requests-html<0.11.0,>=0.10.0->facebook-scraper==0.2.43) (1.4.4)
Requirement already satisfied: zipp>=0.5 in ./.local/lib/python3.6/site-packages (from importlib-metadata<3.0.0,>=2.1.1->pyppeteer>=0.0.14->requests-html<0.11.0,>=0.10.0->facebook-scraper==0.2.43) (3.4.1)
Requirement already satisfied: beautifulsoup4 in /usr/lib/python3/dist-packages (from bs4->requests-html<0.11.0,>=0.10.0->facebook-scraper==0.2.43) (4.6.0)
Requirement already satisfied: cssselect>0.7.9 in ./.local/lib/python3.6/site-packages (from pyquery->requests-html<0.11.0,>=0.10.0->facebook-scraper==0.2.43) (1.1.0)
Requirement already satisfied: lxml>=2.1 in ./.local/lib/python3.6/site-packages (from pyquery->requests-html<0.11.0,>=0.10.0->facebook-scraper==0.2.43) (4.6.3)
Requirement already satisfied: six>=1.5 in ./.local/lib/python3.6/site-packages (from python-dateutil->dateparser<2.0.0,>=1.0.0->facebook-scraper==0.2.43) (1.16.0)

**Reply by Gardek9** [reply]
I know that it hasn't updated because I checked the facebook_scraper.py file which it shows as updated 2 days ago in the repository, but in my folder it shows as last modified on 30 of june.

**Reply by neon-ninja** [reply]
I've released a new version to PyPI, v0.2.44

**Reply by Gardek9** [reply]
And how do I get/install that?

**Reply by neon-ninja** [reply]
`pip3 install -U facebook-scraper`

**Reply by Gardek9** [reply]
Great!!! it's working. I am still running your Nintendo example with 50 pages to see a sample of the output. How do I stop it from running and have it give me the output up that moment?

**Reply by neon-ninja** [reply]
CTRL-C will terminate the program, discarding all output. The code would have to be written to handle KeyboardInterrupt if you still want it to save on CTRL-C

**Reply by Gardek9** [reply]
it wrote the commenters.csv file, but now I am getting:

home/laptop/.local/lib/python3.6/site-packages/facebook_scraper/facebook_scraper.py:388: UserWarning: Facebook served mbasic/noscript content unexpectedly on https://m.facebook.com/photo/view_full_size/?fbid=10112318064148182&ref_component=mbasic_photo_permalink&ref_page=%2Fwap%2Fphoto.php&refid=13&__tn__=%2Cg
  f"Facebook served mbasic/noscript content unexpectedly on {response.url}"
<snip>
  f"Facebook served mbasic/noscript content unexpectedly on {response.url}"
/home/laptop/.local/lib/python3.6/site-packages/facebook_scraper/facebook_scraper.py:388: UserWarning: Facebook served mbasic/noscript content unexpectedly on https://m.facebook.com/photo/view_full_size/?fbid=657233151482838&ref_component=mbasic_photo_permalink&ref_page=%2Fwap%2Fphoto.php&refid=13&__tn__=%2Cg
  f"Facebook served mbasic/noscript content unexpectedly on {response.url}"
  0%|                                     | 28/10318 [03:16<20:33:28,  7.19s/it]TemporarilyBanned, sleeping 10m

**Reply by Gardek9** [reply]
it did NOT write the profiles.csv, yet

**Reply by Gardek9** [reply]
I am getting temporarily banned, sleeping 10m every 30 seconds after a restart. I wanted to get the profiles.csv just to see a sample of the output.

**Reply by Gardek9** [reply]
Now I got a triple banned like so:
TemporarilyBanned, sleeping 10m
TemporarilyBanned, sleeping 10m
TemporarilyBanned, sleeping 10m

**Reply by neon-ninja** [reply]
https://github.com/kevinzg/facebook-scraper/commit/81465b11f7baf01048a14339bdf47623976cb1eb should fix the issue with incorrect noscript detection.

Perhaps requesting so many profiles in such a short amount of time is triggering a ban. You might want to put some more delay in between requesting profiles with something like `time.sleep(10)`

**Reply by neon-ninja** [reply]
https://github.com/kevinzg/facebook-scraper/commit/1d8a7ddd552da5c5a0db9607b3ec6b874b1bc024 Should reduce the amount of requests you need to make, both speeding things up and reducing the likelihood of a temp ban. Use like so:

```python
profile = get_profile(user_id, timeout=60, allow_extra_requests=False)
```

**Reply by neon-ninja** [reply]
You can put these lines:

```python
df = pd.DataFrame(profiles)
df.to_csv("profiles.csv", index=False)
```

in the TemporarilyBanned handler to write out the results in the event of a temp ban

**Reply by neon-ninja** [reply]
Seems I can only get about 50 profiles before I get temp banned

**Reply by Gardek9** [reply]
Ok, I updated the scraper, I don't know where to place the time.sleep(10) line. I placed the df = pd....... lines after the time.sleep(600)

I got a sample of the profile info (I got banned after 15 profiles), but I don't need most of the info there (which means the spider could be lighter). I only need their user id, first name and last name in separate cells if possible, if not that's ok, an email and phone number if there is one on their profile, and I need them to be in the United States (this is key), nothing else (no cover photo, no cover photo url, no work, no education, no relationships, etc.). Their user id and name I get from the commenters.csv file which apparently I can get without getting banned (all bans came while scraping the profiles part).

If I were to only be able to get the commenters.csv (which only gives me user id and name), is there a way to broadcast a message to them just like you would in an email newsletter? If so, how would it be?

**Reply by Gardek9** [reply]
I got banned in the commenters.csv part, too.

**Reply by neon-ninja** [reply]
Why not just post your ad on the page in question?

**Reply by Gardek9** [reply]
I apologize. Perhaps I misunderstood the use of your facebook-scraper. I need to build a laser sharp targeted list to access these people BEFORE I launch my community, the ad would come AFTER the launch. Anything to point me in the right direction would be much appreciated. Thanks in advance. Perhaps an Instagram scraper? I've been trying to find a solution for weeks now.

What about sending a message to the list I got from the commenters.csv? Is there a way to do that without getting banned? I don't know just think out loud....

**Reply by neon-ninja** [reply]
Why not go through commentators manually and message them on Facebook?

**Reply by Gardek9** [reply]
I don't know what that is?

**Reply by neon-ninja** [reply]
When you scroll through the comments section, you can click on people's names to view their profile. On their profile is a button to send them a private message. Why not message them that way?

**Reply by Gardek9** [reply]
I see. It might take some time to contact 100k or 200k users manually which is what I will need to even begin thinking about launching. I thought that automated tools would do that for me, a scraper being one of those tools.

**Reply by Gardek9** [reply]
Like I was asking before, is there a way to send a message to the people from the commenters.csv file? and what about localization (USA)?

**Reply by neon-ninja** [reply]
We're limited by Facebook's rate limits. The things you're asking to do are intentionally difficult to prevent spam. I note you can put vanity URLs in when messaging people, and you can message up to 150 people at a time.

**Reply by Gardek9** [reply]
Ok, is it possible to modify the current spider to pick users from the USA only?

**Reply by neon-ninja** [reply]
In order to do so, you need to get their profile, and check their listed location. But you can only get 50 profiles per 10 minutes it seems. So it would take at least at least a year to get 100K.

**Reply by Gardek9** [reply]
Ok, thank you for your time, I thought that was the whole point of a scraper in the first place (to get thousands of records that are impossible to get manually). I'll keep trying to figure out a solution. Cheers from Punta Del Este, Uruguay!!!

**Comment by ZEESHANalikhantanoli1** [comment]
i have best fb scraper whatspp me +923152122661

**Reply by RameezRaja1** [reply]
Hello,

Using a scraper to collect detailed user info can violate platform policies and lead to bans. For safer methods to grow your followers, check out (https://topfollowerapkss.com/) for tools and strategies.

If you still want to proceed:

Configure the scraper for public data.
Install and set up Tor on Linux Mint 19.3.
Export the results to a CSV file.
Be mindful of legal and ethical considerations.
